hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -881,6 +881,106 @@ def draw_boolean(
         pass
 
 
+class IRTreeLeaf:
+    def __init__(self, *, ir_type: IRTypeName, value: IRType, kwargs: IRKWargsType):
+        self.ir_type = ir_type
+        self.value = value
+        self.kwargs = kwargs
+
+    def _repr_pretty_(self, p, cycle):
+        assert cycle is False
+        p.text(f"{self.ir_type} {self.value} {self.kwargs}")
+
+
+class IRTreeNode:
+    def __init__(
+        self,
+        *,
+        label: int,
+        parent: Optional["IRTreeNode"],
+        index_in_parent: Optional[int] = None,
+    ):
+        self.parent = parent
+        self.index_in_parent = index_in_parent
+        self.label = label
+        self.children: List[Union[IRTreeNode, IRTreeLeaf]] = []
+
+    def _repr_pretty_(self, p, cycle):
+        assert cycle is False
+        p.text(str(self.label))
+        with p.indent(1):
+            for child in self.children:
+                p.break_()
+                p.pretty(child)
+
+
+class IRTree:
+    def __init__(self):
+        # only None when the top example hasn't been started yet
+        self.root: Optional[IRTreeNode] = None
+        self.current_node: Optional[IRTreeNode] = None
+
```

Comment:
I should justify this tree structure. Right now it tracks a tree where the leaves are ir draws and the nodes are examples. We don't actually need this for converting an ir tree into a buffer; tracking just the leaves is sufficient. But I expect this structure to be useful in particular for the shrinker, so I went ahead with the full tree for this pr.

Reply:
Seems reasonable, but we'll need to confirm that performance is OK - the `DataTree` uses a radix tree to save space, and this seems like it could ~duplicate that.